apps/api-extractor/src/generators/ExcerptBuilder.ts

```diff
@@ -0,0 +1,153 @@
+    for (const excerpt of options.embeddedExcerpts || []) {
+      if (excerpt.node) {
+        remainingExcerptNames.add(excerpt.embeddedExcerptName);
+        nameByNode.set(excerpt.node, excerpt.embeddedExcerptName);
+      }
+    }
+
+    const declarationExcerpt: IDeclarationExcerpt = {
+      excerptTokens: [ ],
+      embeddedExcerptsByName: { }
+    };
+
+    ExcerptBuilder._buildSpan(declarationExcerpt, span, {
+      nodeToStopAt: options.nodeToStopAt,
+      nameByNode,
+      remainingExcerptNames,
+      lastReferencedIndex: 0
+    });
+
+    // For any excerpts that we didn't find, add empty entries
+    for (const embeddedExcerptName of remainingExcerptNames) {
+      declarationExcerpt.embeddedExcerptsByName[embeddedExcerptName] = { startIndex: 0, endIndex: 0 };
+    }
+
+    return declarationExcerpt;
+  }
+
+  private static _buildSpan(declarationExcerpt: IDeclarationExcerpt, span: Span,
+    options: IBuildSpanOptions): boolean {
+
+    if (options.nodeToStopAt && span.kind === options.nodeToStopAt) {
+      return false;
+    }
+
+    if (span.kind === ts.SyntaxKind.JSDocComment) {
+      // Discard any comments
+      return true;
+    }
+
+    // Can this node start a excerpt?
+    const embeddedExcerptName: ExcerptName | undefined = options.nameByNode.get(span.node);
+    let excerptStartIndex: number | undefined = undefined;
+    if (embeddedExcerptName) {
+      // Did we not already build this excerpt?
+      if (options.remainingExcerptNames.has(embeddedExcerptName)) {
+        options.remainingExcerptNames.delete(embeddedExcerptName);
+
+        excerptStartIndex = declarationExcerpt.excerptTokens.length;
+        options.lastReferencedIndex = excerptStartIndex;
+      }
+    }
+
+    if (span.prefix) {
+      if (span.kind === ts.SyntaxKind.Identifier) {
+        ExcerptBuilder._appendToken(declarationExcerpt.excerptTokens, ExcerptTokenKind.Reference,
+          span.prefix, options);
+      } else {
+        ExcerptBuilder._appendToken(declarationExcerpt.excerptTokens, ExcerptTokenKind.Content,
+          span.prefix, options);
+      }
+    }
+
+    for (const child of span.children) {
+      if (!this._buildSpan(declarationExcerpt, child, options)) {
+        return false;
+      }
+    }
+
+    if (span.suffix) {
+      ExcerptBuilder._appendToken(declarationExcerpt.excerptTokens, ExcerptTokenKind.Content,
+        span.suffix, options);
+    }
+    if (span.separator) {
+      ExcerptBuilder._appendToken(declarationExcerpt.excerptTokens, ExcerptTokenKind.Content,
+        span.separator, options);
+    }
+
+    // Are we building a excerpt?  If so, add it.
+    if (excerptStartIndex !== undefined) {
+      declarationExcerpt.embeddedExcerptsByName[embeddedExcerptName!] = {
+        startIndex: excerptStartIndex,
+        endIndex: declarationExcerpt.excerptTokens.length
+      };
+
+      options.lastReferencedIndex = declarationExcerpt.excerptTokens.length;
```

Comment:
>ptions.lastReferencedIndex = declarationExcerpt.excerptTokens.length; [](start = 7, length = 69)

We could consider this to be a boolean instead

Reply:
Fixed

---
In reply to: [237279538](https://github.com/Microsoft/web-build-tools/pull/965#discussion_r237279538) [](ancestors = 237279538)